libraries/AP_Scripting/applets/Bendy_ruler.lua

```diff
@@ -0,0 +1,871 @@
+local UPDATE_RATE_HZ = 10
+
+-- MAV_COLLISION_THREAT_LEVEL
+MAV_COLLISION_THREAT_LEVEL = {
+    NONE = 0,         -- Not a threat
+    LOW = 1,          -- Mild concern about this threat
+    HIGH = 2,         -- Craft is panicking and may take action to avoid
+    ENUM_END = 3      -- End of enum
+}
+-- MAV_COLLISION_SRC
+MAV_COLLISION_SRC = {
+    ADSB = 0,                         -- Source is ADSB_VEHICLE packets
+    MAVLINK_GPS_GLOBAL_INT = 1,        -- Source is MAVLink GPS_GLOBAL_INT
+    ENUM_END = 2                      -- End of enum
+}
+MAV_COLLISION_ACTION={
+    NONE=0, -- Ignore any potential collisions 
+    REPORT=1, -- Report potential collision 
+    ASCEND_OR_DESCEND=2, -- Ascend or Descend to avoid threat 
+    MOVE_HORIZONTALLY=3, -- Move horizontally to avoid threat 
+    MOVE_PERPENDICULAR=4, -- Aircraft to move perpendicular to the collision's velocity vector 
+    RTL=5, -- Aircraft to fly directly back to its launch point 
+    HOVER=6, -- Aircraft to stop in place 
+}
+
+
+local MAV_SEVERITY = {EMERGENCY=0, ALERT=1, CRITICAL=2, ERROR=3, WARNING=4, NOTICE=5, INFO=6, DEBUG=7}
+
+local gcs_threat = {
+    src = nil,  
+    src_id = 0,  
+    timestamp_ms = 0,  
+    _location = nil,  
+    _velocity = nil,  
+    threat_level = nil,  
+    closest_approach_xy = 0.0,  
+    closest_approach_z = 0.0,  
+    time_to_closest_approach = 0.0,  
+    distance_to_closest_approach = 0.0  
+}
+
+
+local avoidance_request = {
+    current_loc = nil,
+    target_loc = nil,
+    groundspeed = 0.0,
+    airspeed = 0.0,
+    wind_dir_rad = 0.0,
+    wind_speed = 0.0,
+    request_time_ms = 0
+}
+local avoid_req2 = {
+    current_loc = nil,
+    target_loc = nil,
+    groundspeed = 0.0,
+    airspeed = 0.0,
+    wind_dir_rad = 0.0,
+    wind_speed = 0.0,
+    request_time_ms = 0
+}
+local avoidance_result = {
+    target_loc = nil,
+    new_target_loc = nil,
+    result_time_ms = 0,
+    avoidance_needed = nil
+}
+
+local PARAM_TABLE_KEY = 7
+local PARAM_TABLE_PREFIX = "BENDY_"
```

Comment:
I'm pretty sure we are going to need more than 9 parameters, using "BENDY". I've started using this convention:
prefix = 
Z - identifies a scripting parameter 
XX - scripting function - e.g. PF = Plane Follow - so BR = Bendy Ruler
9 - bank number when you get beyond 9 parameters so

ZBR_ (or maybe ZBR1) for the first bank of 9 bendy ruler parameters
ZBR2_ for the second bank etc.

This means 

- its easy to find _any_ scripted parameters by just searching or "Z" (or Z* in mavproxy).
- the maximum hit for bendy ruler (for up to 99 parameters) is 4 characters.

Which leaves room for coding some reasonable meaning into the remaining characters.

Reply:
Where did you get the idea we can't have more than nine? A table can have 62 or so.

Reply:
I hit this in writing quadplane_terrain_avoid.lua. An error definitely gets thrown if you try to add more than 9 parameters to a table. I think it might be a Lua only restriction.

Reply:
I think its in AP_Param.h - AP_PARAM_MAX_DYNAMIC = 10, even though there is 

```
 if (num_params > 63) {
        return false;
    }

```
at line 2999, I think all the code in add_table() uses AP_PARAM_MAX_DYNAMIC as the limit, and I have found that you definitely can't add more than 10 parameters in a table.